CI Hardware Documentation Creation (#2633)
tock/tock: doc/CI_Hardware.md

# Tock Hardware-in-the-Loop CI
In addition to software tests, Tock allows platform maintainers to deploy CI runners for hardware platforms, to ensure continuous on-metal performance correctness. The hardware CI is federated, and maintainers for each hardware platform are responsible for ensuring that their hardware CI system is set up, accessible, and runs correctly.

The Tock core team maintains the following hardware CI instances:
- Nrf52840dk	UC San Diego; contact: Pat Pannuto <ppannuto@ucsd.edu>, Anthony Quiroga <aquiroga@ucsd.edu> 

> - Hail (WIP)	UC San Diego; contact: Pat Pannuto <ppannuto@ucsd.edu>, Anthony Quiroga <aquiroga@usd.edu>

Reviewer: Can we turn these into links to the boards in the boards folder, i.e. something like `[Nrf52840dk](../boards/nordic/nrf52840dk)` should work

> <!-- npm i -g markdown-toc; markdown-toc -i CI_Hardware.md -->

Reviewer: This comment is actually a command you can/should run in your terminal command line.

`npm` is the Node package manager -- if you don't have node installed, you can follow the directions here to install node: https://nodejs.org/en/download/

once this runs, an actual table of contents should show up between `toc` and `tocstop`

<!-- toc -->
<!-- tocstop -->

![Workflow](tock/doc/images/workflow.png)

# How To Set Up a New Instance

> 

Reviewer: Can we add a leader paragraph here, something to make clear that all of these directions are about setting up the local runner machine. Maybe something like:

> You will need a local machine to run the unit tests that is attached to the Tock development board under test.
>
> This section explains the software setup needed for this box. These directions assume you are using a Raspberry Pi, which will likely be easiest as there are already pins and drivers for GPIO, SPI, and I2C testing available for RPis, but in principle, any Linux-like platform should work.
> 
> **Note:** This machine should be on a stable, reliable network (i.e. a campus or office setup) to ensure it is always available to run tests. Also of note, runners may run arbitrary code from community-generated PRs, which may create a security risk. Once initial configuration is complete, the _only_ network access the runner should need is to Github servers, so this box can be firewalled accordingly.

## Getting Started ([Source](https://goodoomoodoo.github.io/tock-test-harness/GUIDE))

1. Install Ubuntu Server on Raspberry Pi.
    * Follow the guide [here](https://ubuntu.com/tutorials/how-to-install-ubuntu-on-your-raspberry-pi#1-overview) until step 5. We do not need desktop environment.
2. Install the [Requirements](https://github.com/tock/tock/blob/master/doc/Getting_Started.md#requirements) here.
3. Download and install JLinkExe
```bash
$ wget --post-data 'accept_license_agreement=accepted&non_emb_ctr=confirmed&submit=Download+software' https://www.segger.com/downloads/jlink/JLink_Linux_arm64.tgz
$ tar xvf JLink_Linux_arm64.tgz
$ sudo cp JLink_Linux_V700a_arm64/99-jlink.rules /etc/udev/rules.d/ # Depends on JLink version
 

> # Add the Jlink directory to the path in .profile or .bashrc

Reviewer: Can you add an example command that shows how to do this?

```
4. Install pip3, if not yet installed
```bash
sudo apt install python-pip3
```
5. Checkout test harness in home directory 
```bash
cd ~; git clone https://github.com/goodoomoodoo/tock-test-harness.git
cd ~; git clone https://github.com/goodoomoodoo/libtock-c.git
```
6. Install gpiozero if not yet installed. (It should come with the image)
```bash
$ sudo apt install python3-gpiozero
$ sudo chown root:$USER /dev/gpiomem
$ sudo chmod g+rw /dev/gpiomem
```
7. Install linux library
```bash
$ sudo apt install arm-none-eabi
# Install bluetooth library
$ sudo apt-get install bluez bluez-tools pi-bluetooth 

> # !IMPORTANT! Remember to reboot after installation

Reviewer: ```suggestion
# !IMPORTANT! Remember to reboot after installation
$ sudo reboot
```

3. Download and install JLinkExe
```bash
$ wget --post-data 'accept_license_agreement=accepted&non_emb_ctr=confirmed&submit=Download+software' https://www.segger.com/downloads/jlink/JLink_Linux_arm64.tgz
$ tar xvf JLink_Linux_arm64.tgz
$ sudo cp JLink_Linux_V700a_arm64/99-jlink.rules /etc/udev/rules.d/ # Depends on JLink version
 
# Add the Jlink directory to the path in .profile or .bashrc
```
4. Install pip3, if not yet installed
```bash
sudo apt install python-pip3
```
5. Checkout test harness in home directory 
```bash
cd ~; git clone https://github.com/goodoomoodoo/tock-test-harness.git
cd ~; git clone https://github.com/goodoomoodoo/libtock-c.git
```
6. Install gpiozero if not yet installed. (It should come with the image)
```bash
$ sudo apt install python3-gpiozero
$ sudo chown root:$USER /dev/gpiomem
$ sudo chmod g+rw /dev/gpiomem
```
7. Install linux library
```bash
$ sudo apt install arm-none-eabi
# Install bluetooth library
$ sudo apt-get install bluez bluez-tools pi-bluetooth 
# !IMPORTANT! Remember to reboot after installation
```
8. Install python3 libraries
```bash
$ cd ~/tock-test-harness; pip3 install -r requirements.txt
```

> 9. Hook up Action Runner of the Tock Repo to the Raspberry Pi.

Reviewer: Add a note that this step will have to be done by a member of `@tock/core-wg`

    * Go to \'Settings\' and go to \'Actions\' down the list on the left
    * Under \'Self-hosted runners\' click **Add runner**
    * Follow the steps to set up action runner on Raspberry Pi

10. Add path to JLink to ```.path``` in Action Runner
```bash
$ cd ~/actions-runner; vi .path

> # Add the path

Reviewer: Since we are controlling the whole setup here, can we just give the exact command. In fact, can we skip the editor step, replace this with something like:

`echo "JLINK_PATH_WHATEVER_IT_IS" > ~/actions-runner/.path`

?

```sudo netplan generate; sudo netplan apply```

### Action runner: **command not found**

If you experience issue with the command not being found, but you are very sure that you have installed the corresponding software. Try to uninstall the runner and reinstall it.

```bash
# 1. Uninstall Action Runner Server
$ cd ~/actions-runner
$ sudo ./svc.sh stop
$ sudo ./svc.sh uninstall
$ ./config.sh remove

# 2. Reinstall Action Runner Server
# Follow the guide in Getting Started.
```
**Note**: Possible reason to this issue is that the server environment and path varaible cannot be update after it has started. Thus, it requires a reconfiguration to included the updated path and environment variables.

## Creating the Configuration File
After all is set up, the first step in creating an instance is to run the [runner_init](https://github.com/goodoomoodoo/tock-test-harness/blob/main/runner_init.py) script to create a configuration file for the raspberry pi you just set up.
This configuration file will then be used in the workflow of actions for github whenever an update is pushed to be accessed by the runners, which will be denoted as config.toml for simplicity sake here.

> 

Reviewer: just noting the todo we already talked about: add example of config file here

- Communication Protocol 
    - Used to run tockloader
    - The list of communication protocols are provided, and one chosen must be in the list. 
        - Typically Jlink is the one chosen here

When runner_init script is executed, there will be prompts to fill in this information and the configuration file will be created based on the responses, filling in the file with the information listed above.

# How Instances Work
Instances in this case are workflows with Github actions that set runners to complete said actions. This is how testing is conducted in Hardware Continuous Integration. The entire process is taken within the workflow through the yml files, which are another type of configuration file. Currently using the tock-hw-ci.yml as the workflow to run tests, in which we will use as an example throughout this document. (This file is located [here](https://github.com/goodoomoodoo/tock/blob/master/.github/workflows/tock-hw-ci.yml))

## Process of Instances
As stated instances work through GitHub actions that set up runners. These are called workflows, and are set to call whenever a specific action occurs.

### What are actions/runners?

> Before getting into the 

Reviewer: delete?

This is the workflow of how instances work through Hardware CI. Ideally, every board will have this workflow to be called whenever updates occur and follow this form processing. 

# How Tests are Chosen
Tests are chosen when initially running the program through the setup/configuration of the Raspberry Pi, where you build a configuration file that is run through the runners and when creating the file. Essentially, when setting up the Raspberry Pi, you create a configuration file on boot up with the get started guide listed that sets the specified board type, and that board type will have a test configuration file containing the tests for all boards of that type and/or the specific Raspberry Pi through the harness Identity number. Then, tests are chosen for the specified board that is supported by the Tock OS.
As noted in the previous section we have a test.config.toml file that contains information of the test to run for each board in the Tock/Boards directory , which contains each specific Tock supported board as well. Thus to Add tests, there are two primary steps:
- **Adding Tests**

>     - **1** To add tests, you’ll need to create two files: a main.c (c) file and a test.py (python) file.

Reviewer: Reformat this to be a markdown-style numbered list.

![Diagram for Adding/Choosing Tests](tock/doc/images/testsdiagram.png)

# What to do if Something Goes Wrong

## If a test fails
Take a look at what type of test failed (in the output on github) and see what is the issue of the build of the board accordingly. The output in github currently prints what test failed to give the user a better understanding of what’s wrong in the installation.

## If there’s boot up issues of the Raspberry Pi

> Go to Troubleshooting section [here.](#troubleshoot)

Reviewer: ```suggestion
Go to the [Troubleshooting section](#troubleshoot).
```

Modern accessibility guidelines say to avoid `here` links, as screen readers do not handle them well.

- Board Path
    - This is the path of the Tock supported board in the official tock repository. The path must be provided when setting up the configuration. This is essentially the directory relative to the Tock boards directory. 
- Harness ID
    - This is the specific Identity Number of the Raspberry Pi. This will be used in the test.config.toml which is kept for each board type to run tests on in the runner. 
- Communication Protocol 
    - Used to run tockloader
    - The list of communication protocols are provided, and one chosen must be in the list. 
        - Typically Jlink is the one chosen here

When runner_init script is executed, there will be prompts to fill in this information and the configuration file will be created based on the responses, filling in the file with the information listed above.

**Example Configuration File:**

```toml
title = "pi_test"

[env]
board = "nrf52dk"
path = "{home}/actions-runner/_work/tock/tock/boards/nordic/nrf52dk/"
harness_id = "{Pi ID}"
communication_protocol = "jlink"
```

> **Note:** {home} and {Pi ID} would depend on your configuration, where {home} will be your home directory to start the path, and {Pi ID} will be your specific Raspberry Pi ID

Reviewer: Where is Pi ID assigned? Is that something set on the RPi or is it something set in GitHub somewhere?

```

## Downloading Pi-GPIO open source Python Library

> [Source of Guide](https://abyz.me.uk/rpi/pigpio/download.html)

Reviewer: ```suggestion
For more details or troubleshooting, [see the complete pigpio documentation](https://abyz.me.uk/rpi/pigpio/download.html).
```
